// @start9labs/start-sdk/package/lib/util/Drop.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Drop = void 0;
class Drop {
    constructor() {
        this.dropId = Drop.idCtr++;
        this.dropRef = { id: this.dropId };
        const weak = this.weak();
        Drop.weak[this.dropId] = weak;
        Drop.registry.register(this.dropRef, this.dropId, this.dropRef);
        return new Proxy(this, {
            set(target, prop, value) {
                if (prop === "dropRef" || prop == "dropId")
                    return false;
                target[prop] = value;
                weak[prop] = value;
                return true;
            },
        });
    }
    register() { }
    weak() {
        const weak = Object.assign(Object.create(Object.getPrototypeOf(this)), this);
        if (this.dropRef)
            weak.ref = new WeakRef(this.dropRef);
        return weak;
    }
    drop() {
        if (!this.dropRef || !this.dropId)
            return;
        this.onDrop();
        this.leak();
    }
    leak() {
        if (!this.dropRef || !this.dropId)
            return this;
        Drop.registry.unregister(this.dropRef);
        delete Drop.weak[this.dropId];
        delete this.dropRef;
        delete this.dropId;
        return this;
    }
}
exports.Drop = Drop;
Drop.weak = {};
Drop.registry = new FinalizationRegistry((id) => {
    const weak = Drop.weak[id];
    if (weak)
        weak.drop();
});
Drop.idCtr = 0;
