// @start9labs/start-sdk/package/lib/inits/setupInstall.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PostInstall = exports.PreInstall = exports.Install = void 0;
exports.setupPreInstall = setupPreInstall;
exports.setupPostInstall = setupPostInstall;
class Install {
    constructor(fn) {
        this.fn = fn;
    }
}
exports.Install = Install;
class PreInstall extends Install {
    constructor(fn) {
        super(fn);
    }
    static of(fn) {
        return new PreInstall(fn);
    }
    async preInstall({ effects }) {
        await this.fn({
            effects,
        });
    }
}
exports.PreInstall = PreInstall;
function setupPreInstall(fn) {
    return PreInstall.of(fn);
}
class PostInstall extends Install {
    constructor(fn) {
        super(fn);
    }
    static of(fn) {
        return new PostInstall(fn);
    }
    async postInstall({ effects }) {
        await this.fn({
            effects,
        });
    }
}
exports.PostInstall = PostInstall;
function setupPostInstall(fn) {
    return PostInstall.of(fn);
}
