// @start9labs/start-sdk/base/lib/util/getServiceInterfaces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetServiceInterfaces = void 0;
exports.getServiceInterfaces = getServiceInterfaces;
const getServiceInterface_1 = require("./getServiceInterface");
const makeManyInterfaceFilled = async ({ effects, packageId, callback, }) => {
    const serviceInterfaceValues = await effects.listServiceInterfaces({
        packageId,
        callback,
    });
    const serviceInterfacesFilled = await Promise.all(Object.values(serviceInterfaceValues).map(async (serviceInterfaceValue) => {
        const hostId = serviceInterfaceValue.addressInfo.hostId;
        const host = await effects.getHostInfo({
            packageId,
            hostId,
            callback,
        });
        if (!host) {
            throw new Error(`host ${hostId} not found!`);
        }
        return {
            ...serviceInterfaceValue,
            host,
            addressInfo: (0, getServiceInterface_1.filledAddress)(host, serviceInterfaceValue.addressInfo),
        };
    }));
    return serviceInterfacesFilled;
};
class GetServiceInterfaces {
    constructor(effects, opts) {
        this.effects = effects;
        this.opts = opts;
    }
    /**
     * Returns the service interfaces for the package. Reruns the context from which it has been called if the underlying value changes
     */
    async const() {
        const { packageId } = this.opts;
        const callback = this.effects.constRetry &&
            (() => this.effects.constRetry && this.effects.constRetry());
        const interfaceFilled = await makeManyInterfaceFilled({
            effects: this.effects,
            packageId,
            callback,
        });
        return interfaceFilled;
    }
    /**
     * Returns the service interfaces for the package. Does nothing if the value changes
     */
    async once() {
        const { packageId } = this.opts;
        const interfaceFilled = await makeManyInterfaceFilled({
            effects: this.effects,
            packageId,
        });
        return interfaceFilled;
    }
    /**
     * Watches the service interfaces for the package. Returns an async iterator that yields whenever the value changes
     */
    async *watch() {
        const { packageId } = this.opts;
        const resolveCell = { resolve: () => { } };
        this.effects.onLeaveContext(() => {
            resolveCell.resolve();
        });
        while (this.effects.isInContext) {
            let callback = () => { };
            const waitForNext = new Promise((resolve) => {
                callback = resolve;
                resolveCell.resolve = resolve;
            });
            yield await makeManyInterfaceFilled({
                effects: this.effects,
                packageId,
                callback,
            });
            await waitForNext;
        }
    }
    /**
     * Watches the service interfaces for the package. Takes a custom callback function to run whenever the value changes
     */
    onChange(callback) {
        ;
        (async () => {
            for await (const value of this.watch()) {
                try {
                    await callback(value);
                }
                catch (e) {
                    console.error("callback function threw an error @ GetServiceInterfaces.onChange", e);
                }
            }
        })()
            .catch((e) => callback(null, e))
            .catch((e) => console.error("callback function threw an error @ GetServiceInterfaces.onChange", e));
    }
    /**
     * Watches the service interfaces for the package. Returns when the predicate is true
     */
    async waitFor(pred) {
        const { packageId } = this.opts;
        const resolveCell = { resolve: () => { } };
        this.effects.onLeaveContext(() => {
            resolveCell.resolve();
        });
        while (this.effects.isInContext) {
            let callback = () => { };
            const waitForNext = new Promise((resolve) => {
                callback = resolve;
                resolveCell.resolve = resolve;
            });
            const res = await makeManyInterfaceFilled({
                effects: this.effects,
                packageId,
                callback,
            });
            if (pred(res)) {
                resolveCell.resolve();
                return res;
            }
            await waitForNext;
        }
        return null;
    }
}
exports.GetServiceInterfaces = GetServiceInterfaces;
function getServiceInterfaces(effects, opts) {
    return new GetServiceInterfaces(effects, opts);
}
